solidago/src/solidago/mehestan/individual/continuous_bradly_terry.py

```diff
@@ -0,0 +1,156 @@
+        np.where(
+            np.abs(theta_ab) < EPSILON,
+            theta_ab / 3,
+            1 / np.tanh(theta_ab) - 1 / theta_ab,
+        )
+        + r_ab
+    )
+
+
+@njit
+def Delta_theta(theta_ab):
+    return np.where(
+        np.abs(theta_ab) < EPSILON,
+        1 / 3 - 1 / 15 * theta_ab**2,
+        1 - (1 / np.tanh(theta_ab)) ** 2 + 1 / theta_ab**2,
+    ).sum() ** (-0.5)
+
+
+@njit
+def coordinate_optimize(r_ab, theta_b, precision, alpha):
+    theta_low = -5.0
+    while contributor_loss_partial_derivative(theta_low, theta_b, r_ab, alpha) > 0:
+        theta_low *= 2
+
+    theta_up = 5.0
+    while contributor_loss_partial_derivative(theta_up, theta_b, r_ab, alpha) < 0:
+        theta_up *= 2
+
+    return brentq(
+        contributor_loss_partial_derivative,
+        theta_low,
+        theta_up,
+        args=(theta_b, r_ab, alpha),
+        xtol=precision,
+    )
+
+
+class ContinuousBradleyTerry(IndividualScoresAlgorithm):
+    def __init__(self, r_max, alpha=DEFAULT_ALPHA):
+        """
+        Initialize Continous Bradley Terry (CBT) to compute individual scores
+
+        Parameters
+        ----------
+        r_max: Maximum absolute score of a comparison
+        alpha: Regularization term
+        """
+        self.alpha = alpha
+        self.r_max = r_max
+
+    def coordinate_descent(self, coord_to_subset, initial_scores):
+        n_alternatives = len(coord_to_subset)
+        unchanged = set()
+        to_pick = []
+
+        def pick_next_coordinate():
+            nonlocal to_pick
+            if len(to_pick) == 0:
+                to_pick = list(range(n_alternatives))
+                random.shuffle(to_pick)
+            return to_pick.pop()
+
+        theta = initial_scores
+        while len(unchanged) < n_alternatives:
+            coord = pick_next_coordinate()
+            if coord in unchanged:
+                continue
+            indices, r_ab = coord_to_subset[coord]
+            old_theta_a = theta[coord]
+            theta_b = theta[indices]
+            new_theta_a = coordinate_optimize(
+                r_ab, theta_b, precision=EPSILON / 10, alpha=self.alpha
+            )
+            theta[coord] = new_theta_a
+            if abs(new_theta_a - old_theta_a) < EPSILON:
+                unchanged.add(coord)
+            else:
+                unchanged.clear()
+        return theta
+
+    def compute_individual_scores(self, scores: pd.DataFrame, initial_entity_scores=None):
+        scores = scores[["entity_a", "entity_b", "score"]]
+        scores_sym = pd.concat(
+            [
+                scores,
+                pd.DataFrame(
+                    {
+                        "entity_a": scores.entity_b,
+                        "entity_b": scores.entity_a,
+                        "score": -1 * scores.score,
+                    }
+                ),
+            ]
+        )
+        r = scores_sym.pivot(index="entity_a", columns="entity_b", values="score") / self.r_max
```

Comment:
Same comment

Reply:
This `.pivot()` call has been removed from CBT in  3f48497